### ISSUE TITLE
cmd/gc: x op= y is not always equivalent to x = x op y.

### ISSUE THREAD

**Issue by cznic** [golang/go#11161]
From the [specs](http://golang.org/ref/spec#Assignments)

> An *assignment operation x op= y* where *op* is a binary arithmetic operation is equivalent to *x = x op y* but evaluates `x` only once.

This [program](http://play.golang.org/p/1VZyzDnjND)

```go
package main

import "fmt"

func main() {
	a := 0.5
	a = a * 2 / 3
	fmt.Println(a)

	b := 0.5
	b *= 2 / 3
	fmt.Println(b)
}
```

Outputs

```
0.3333333333333333
0
```

(Issue originally submited by @crmackay as #11158, but closed and locked, hence resubmitting here.)

**Comment by champioj** [comment]
The spec say: 
>Binary operators of the same precedence associate from left to right. For instance, x / y * z is the same as (x / y) * z.

So one is  `(a*2)/3` and the other is `a*(2/3)`
While it is surprising,  I think this is the correct behavior.

    func main() {
        a := 0.5
        a = a * 2 + 3
        fmt.Println(a)

        b := 0.5
        b *= 2 + 3
        fmt.Println(b)
    }
This should not print the same value. If you think the problem is elsewhere, it would be good to be more specific.

**Comment by cznic** [comment]
@champioj I see and I am confused (and probably wrong).

It surprises me that this [program](http://play.golang.org/p/QR23c2WhOG)

```go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(1.234 * 5 / 7)
	fmt.Println(5 / 7 * 1.234)
}
```

outputs

```
0.8814285714285715
0
```

From the [specs](http://golang.org/ref/spec#Constant_expressions):

> Constant expressions are always evaluated exactly; intermediate values and the constants themselves may require precision significantly larger than supported by any predeclared type in the language. 

But the reason seems to be

> Except for shift operations, if the operands of a binary operation are different kinds of untyped constants, the operation and, for non-boolean operations, the result use the kind that appears later in this list: integer, rune, floating-point, complex.

With the explicit specs example

> `const b = 15 / 4           // b == 3     (untyped integer constant)`

**Comment by griesemer** [comment]
@cznic Your observation is correct. If you change your program to http://play.golang.org/p/PxBAbmlIiU you get the same result in both cases:

1.234 * 5 / 7 is the same as (1.234 * 5) / 7, where / is a floating-point division (resulting in a fractional value)

5 / 7 * 1.234 is the same as (5 / 7) * 1.234, where / is an integer division (since 5, 7 are integer literals), resulting in 0.

If you make the 5 a float (5. or 5.0) you get the same results on both lines.

Working as intended.

**Comment by mdempsky** [comment]
I agree that gc's current behavior of treating "x *= 2 / 3" like "x = x * (2 / 3)" seems correct/intuitive, but does the spec actually explain that unambiguously?  I don't see anything that rules out @cznic's expectation that it should have instead behaved like "x = (x * 2) / 3".

The simplest fix seems to be to clarify that "x op= y" is equivalent to "x = x op (y)".

**Comment by griesemer** [comment]
@mdempsky The English wording is not to be taken as a "macro expansion" of y but I agree it's unclear. Change pending. Thanks for pointing this out.

**Comment by cznic** [comment]
@mdempsky That's a nice suggestion and I see @griesemer has already implemented it. Thanks to both of you.

A side note: What tripped me in the understanding of this particular part of specs is this

> Constant expressions are always evaluated exactly; ...

I just stopped reading after the word *exactly*, as I automatically, but wrongly, assumed it naturaly implies arithmetic commutativity/associativity is granted for the *whole expression* - as is the case with *exact*/symbolic arithmetics, like seen in Maxima or Mathematica or any other CAS.

I learned that actually the specified exactness works granularly - for every [binary] operation in particular, but then the intermediate result is possibly turned *inexact* by applying a specific domain, eg. ℤ - with the loss of arithmetic commutativity and/or associativity.

I can see the reason for this difference wrt to "ideal" const expression evaluation and I'm fine with it.

Thanks again to all for the explanation.
